File: src/main/java/com/github/androidutils/logger/StartupLogWriter.java
package com.github.androidutils.logger;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.concurrent.CopyOnWriteArrayList;

import com.github.androidutils.logger.Logger.LogLevel;
import com.github.androidutils.logger.Logger.LogWriter;

public class StartupLogWriter implements LogWriter {

    private static final int STARTUP_BUFFER_SIZE = 100;

    private final DateFormat dtf;
    private final CopyOnWriteArrayList<String> messages;

    private StartupLogWriter() {
        dtf = new SimpleDateFormat("dd-MM HH:mm:ss");
        messages = new CopyOnWriteArrayList<String>();
    }

    public CopyOnWriteArrayList<String> getMessages() {
        return messages;
    }

    public String getMessagesAsString() {
        StringBuilder sb = new StringBuilder(messages.size() * 120);
        for (String message : messages) {
            sb.append(message).append('\n');
        }
        return sb.toString();
    }

    @Override
    public void write(LogLevel level, String tag, String message) {
        write(level, tag, message, null);
    }

    @Override
    public void write(LogLevel level, String tag, String message, Throwable throwable) {
        if (messages.size() < STARTUP_BUFFER_SIZE) {
            // BufferedWriter for performance, true to set append to file flag
            final StringBuilder buf = new StringBuilder();
            final Date timeStamp = new Date(System.currentTimeMillis());
            buf.append(dtf.format(timeStamp));
            buf.append(" ");
            buf.append(level.name());
            buf.append(" ");
            buf.append(tag);
            buf.append(" ");
            buf.append(message);
            if (throwable != null) {
                buf.append(throwable.toString());
            }
            messages.add(buf.toString());
        }
    }

    private static volatile StartupLogWriter sInstance;

    public static synchronized StartupLogWriter getInstance() {
        if (sInstance == null) {
            sInstance = new StartupLogWriter();
        }
        return sInstance;
    }
}

File: app/src/main/java/com/better/alarm/logger/StartupLogWriter.kt
package com.better.alarm.logger

import androidx.collection.CircularArray
import com.better.alarm.logger.Logger.LogLevel
import java.text.DateFormat
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

class StartupLogWriter private constructor() : LogWriter {
    private val bufferSize = 1024
    private val dtf: DateFormat = SimpleDateFormat("dd-MM HH:mm:ss", Locale.GERMANY)

    /**
     * guarded by [lock], because of the ContentProvider
     */
    private val messages = CircularArray<String>()
    private val lock = ReentrantLock()

    fun getMessagesAsString(): String {
        return lock.withLock {
            generateSequence { if (messages.isEmpty) null else messages.popFirst() }
                    .joinToString("\n")
        }
    }

    override fun write(level: LogLevel, tag: String, message: String, throwable: Throwable?) {
        val timeStamp = Date(System.currentTimeMillis())
        val exceptionMessage = throwable?.message ?: ""

        lock.withLock {
            if (messages.size() == bufferSize) {
                messages.popFirst()
            }
            messages.addLast("${dtf.format(timeStamp)} ${level.name} $tag $message $exceptionMessage")
        }
    }

    companion object {
        @JvmStatic
        fun create(): StartupLogWriter {
            return StartupLogWriter()
        }
    }
}
